### DISCUSSION TITLE
Conditional on feed from Google Calendar

### DISCUSSION THREAD

**Discussion by danielvivacqua** [engram-design/FeedMe#896] [Ideas]
### Description

I am working on importing Google Calendar events with Feed Me. The main problem we're running into is importing all day events correctly. Entries are being created with post date and expiry date as the Google Calendar events start/end date/times.

However, the Google Calendar JSON uses either start.dateTime for timed events (e.g, 1:00pm – 3:00pm) OR start.date for all-day events.

If the import doesn't find the thing that post/expiry dates are mapped to, it assuming the date/time the entry is imported.

Ideally, we'd like to be able to import feeds with a condition like "if start.dateTime exists, import that as the post date, otherwise use start.date and 12:00am"

### Steps to reproduce

1. Using Feed Me, import a Google Calendar in the Feed URL using the Google Calendar API such as https://www.googleapis.com/calendar/v3/calendars/[url redacted for client privacy]timeMin=2019-04-01T00:00:00-07:00
2. For Feed Type, choose JSON.
3. For Element Type, choose Entry.
4. Choose the appropriate Section (in this case, the specific Calendar section)
5. For Entry Type, choose the entry you have created for a Calendar event.
6. For Import strategy, choose "Create new elements", "Update existing elements", and "Disable missing elements"
7. Hit "Save & Continue"
8. The Primary Element field will say /items (xNumber elements). Pagination URL should be No Pagination URL. Hit "Save & Continue"
9. For Title, choose (from the JSON) `<summary>`. Do not set a Default Value.
10. For Slug, Don't import
11. For Post Date *and this is where we need the conditional* choose `<start/dateTime>` from the JSON. NOTE: Ideally, we would like to make it so that a Google calendar event that is all day uses `<start/date>` and automatically gets the time set to "12:00AM", as we have a conditional in the twig that says `{% if dateTime == "12:00AM" %} <p>All Day</p> {% endif %}`
12. For Expiry Date *(also where we need the conditional)* choose `<end/dateTime>`
12. Status, Don't import. Author, Don't import. Entry ID, Don't import.
13. Add additional fields to your Calendar entry type, fields Event ID (plain text) and Event Link (plain text). For Event ID, choose `<id>` from the Google calendar JSON. For Event Link, choose `<htmlLink>` from the JSON.
14. Under "Set a unique identifier to match against existing elements", choose Event ID.
15. Save & Continue
16. The feed should be ready to run, but events that are in the Google calendar as All Day events will create entries that have the Post Date of when you ran the feed and the Expiry Date of now. This prevents the client from being able to post all day events on the correct date.

### Additional info

- Plugin version: 4.0.0
- Craft version: Craft Pro 3.1.22

**Comment by brandonkelly** [comment]
Conditional field mapping is an interesting idea!

For now, I think the only way to do this would be to split out your all-day events into a separate JSON file, and create a separate feed in Feed Me for that.

**Reply by gixxy22** [reply]
Conditionals would be amazing, only import if feed data is 'something' than 'something'

Only update if feed data is 'something' than entry data.

Combined with twig, the possibilities are limitless.

**Comment by daltonrooney** [comment]
Obvioiusly too late to be helpful for OP, but you should be able to do this in the `stepBeforeElementSave` event.

```
Event::on(Process::class,Process::EVENT_STEP_BEFORE_ELEMENT_SAVE,function(FeedProcessEvent $event){
  
/* Pseudocode 
  $event->feedData contains an array of the parsed feed content from the node
  $event->element contains the element to be saved
 
  Conditional logic:  
  if ( $event->feedData['startDate'] == "something" ) {
   $event->element->startDate = "something"
  } else {
    $event->element->startDate = "something else"
  }
*/

});
```
